Updated Pull Request for CloudNative Ch1... (#13709)
dotnet/docs: docs/architecture/cloud-native/definition.md

- Each is self-contained. The domain logic, data, and state are encapsulated in the service. You select the data storage technology (SQL, NoSQL) and programming platform of choice.

- Each is developed autonomously and can be deployed independently.

- Each runs in its own process and communicates with others using standard communication protocols such as HTTP/HTTPS, WebSockets, or [AMQP](https://en.wikipedia.org/wiki/Advanced_Message_Queuing_Protocol).

With microservices, you move away from a traditional monolithic design, where an entire application executes in a single-tiered environment. Figure 1-4 contrasts a monolithic and microservices approach. 

![Monolithic deployment versus microservices](media/monolithic-deployment-vs-microservices.jpg)

**Figure 1-4.** Monolithic deployment versus microservices

### Why microservices?

Microservices provide agility.

Each microservice is independent with an autonomous lifecycle. It can evolve independently and deploy frequently. You don’t have to wait for a quarterly release to deploy new features or updates. You can update small areas of a complex application with less risk of disrupting the entire system.

> Each microservice can scale independently. Instead of scaling the entire application as a single unit, you scale out only those services that require more processing power or network bandwidth. This  fine-grained approach to scaling provides for greater control of your system and helps to reduce overall costs as you scale portions of your system, not everything.

Reviewer: ```suggestion
Each microservice can scale independently. Instead of scaling the entire application as a single unit, you scale out only those services that require more processing power or network bandwidth. This fine-grained approach to scaling provides for greater control of your system and helps to reduce overall costs as you scale portions of your system, not everything.
```

Each microservice can scale independently. Instead of scaling the entire application as a single unit, you scale out specific microservices that require more processing power or network bandwidth. Such fine-grained scaling provides greater control of your system and helps to reduce overall costs as you scale portions of your system, not everything.

> An excellent reference guide for microservices is [.NET Microservices: Architecture for Containerized .NET Applications](https://docs.microsoft.com/dotnet/standard/microservices-architecture/).

Reviewer 2: ```suggestion
An excellent reference guide for microservices is [.NET Microservices: Architecture for Containerized .NET Applications](../microservices/index.md).
```

## Deployed in containers…

> Microservices go hand in hand with software containerization. Developers package the code, dependencies, and runtime for a service as a single unit, called an [image](https://docs.docker.com/glossary/?term=image). Images are stored in a [container registry](https://azure.microsoft.com/services/container-registry/). Registries can be located on your laptop, in your data center, or in the public cloud. When needed, you *pull* the image and transform it into a running container instance. The instance runs on any computer (development laptop, QA, or production) that has the Docker software engine.

Reviewer 2: ```suggestion
Microservices go hand in hand with software containerization. Developers package the code, dependencies, and runtime for a service as a single unit, called an [image](https://docs.docker.com/glossary/?term=image). Images are stored in a [Container Registry](https://azure.microsoft.com/services/container-registry/). Registries can be located on your laptop, in your data center, or in the public cloud. When needed, you *pull* the image and transform it into a running container instance. The instance runs on any computer (development laptop, QA, or production) that has the Docker software engine.
```

### Developing in containers

Most developers build code on their local computers. When complete, the code is unit tested and checked into a team repository for future use. 

Microservice systems, however, are often large and perform poorly on local development machines. Many development teams containerize their cloud-native development cycle. They write code in Docker containers and run the application in a shared development cluster in the cloud. 

 [Azure Dev Spaces](https://docs.microsoft.com/azure/dev-spaces/) enables development teams to iteratively build and test an entire microservices application running in [Azure Kubernetes Service](https://azure.microsoft.com/services/kubernetes-service/) (AKS). Azure Dev Spaces is shown in Figure 1-6.

![Azure Dev Spaces](media/azure-dev-spaces-site-example.png)

**Figure 1-6**. Azure Dev Spaces

> In the previous figure, Developer Susie Walker and team share a development instance of their entire application in an [Azure Kubernetes Service](https://azure.microsoft.com/services/kubernetes-service/) cluster (in blue). Using Azure Dev Spaces, Susie creates a personal development space with a private copy of the Hotel Service, shown in gray. She can then work with her copy of the hotel service, but seamlessly run it against the entire application in the AKS development cluster. This way, Suzie doesn't need to install the full code base or have to mock or replicate dependencies on her local computer. Until Susie deploys her updated copy of the Hotel Service to the dev cluster, other team members run against the older version in the cluster. We discuss this upcoming product, Azure Dev Spaces, later in this book.

Reviewer: ```suggestion
In the previous figure, developer Susie Walker and team share a development instance of their entire application in an [Azure Kubernetes Service](https://azure.microsoft.com/services/kubernetes-service/) cluster (in blue). Using Azure Dev Spaces, Susie creates a personal development space with a private copy of the Hotel Service, shown in gray. She can then work with her copy of the hotel service, but seamlessly run it against the entire application in the AKS development cluster. This way, Suzie doesn't need to install the full code base or have to mock or replicate dependencies on her local computer. Until Susie deploys her updated copy of the Hotel Service to the dev cluster, other team members run against the older version in the cluster. We discuss this upcoming product, Azure Dev Spaces, later in this book.
```

![Cloud-native backing services](media/cloud-native-backing-services.png)

**Figure 1-7**. Cloud-native backing services

Cloud-native applications are best built using *cloud hosted backing services* managed by a cloud provider. The provider operates these resources at scale and bears the responsibility for security and maintenance. Your services consume each backing service from a RESTFul endpoint.

Built on cloud technologies, managed backing services are more agile and flexible than traditional offerings. They offer inherent scalability and resiliency and are continually optimized for speed, security, and performance.

Importantly, cloud providers fully support their managed backing services. Open a ticket and they fix your issue. 

Instead, your operations team could accept the responsibility for owning a backing service. They could provision a VM and install the open-source RabbitMQ message broker. While feasible, would this make sense?

> Chances are you’re not in the message broker business. So, why incur the responsibility for owning this service when the cloud provider can do it better and for less? No matter how skilled you are, it's unlikely that you can match the reliability and performance of [Azure Service Bus](https://azure.microsoft.com/services/service-bus/), a fully managed message broker service in Azure. The team working on Azure Service Bus specializes in message broker technology - that’s their business. While you, instead, sell consumer items, provide billing services, or deliver packages. Do what you're good at and let others help you with the rest. Focus your effort on building outstanding customer functionality, not becoming experts in open-source backing services.

Reviewer: ```suggestion
Chances are you’re not in the message broker business. So why incur the responsibility for owning this service when the cloud provider can do it better and for less? No matter how skilled you are, it's unlikely that you can match the reliability and performance of [Azure Service Bus](https://azure.microsoft.com/services/service-bus/), a fully managed message broker service in Azure. The team working on Azure Service Bus specializes in message broker technology - that’s their business - while you, instead, sell consumer items, provide billing services, or deliver packages. Do what you're good at and let others help you with the rest. Focus your effort on building outstanding customer functionality, not becoming experts in open-source backing services.
```

## Embrace modern design…

How would you design a cloud-native app? What would your architecture look like? To what principles, patterns, and best practices would you adhere? So far we’ve talked about cloud compute, microservices, containers, and managed backing services. But, what about the application design process itself?

We start with a widely accepted foundation for service design and then highlight some important design concerns.

### The 12-Factor Foundation

A widely accepted methodology for constructing cloud-native systems is the [Twelve-Factor Application](https://12factor.net/). It describes guidelines for building distributed cloud-based applications. Many practitioners consider it a foundation for building cloud-native apps.

Figure 1-9 shows the 12-Factor methodology.

![The 12-Factor app](media/the-twelve-factor-app.png)
**Figure 1-9.** The 12-Factor app

We'll refer to these 12 factors throughout the book.

### Critical Design Considerations

Beyond the guidance provided from the 12-factor manifesto, there are several critical design principles you must consider.

*Communication*

> How will front-end client applications communicate with your backed-end core services? Will you allow direct communication? Or, might you abstract the backend services with a gateway façade that provides  flexibility, control, and security?

Reviewer: ```suggestion
How will front-end client applications communicate with your backed-end core services? Will you allow direct communication? Or might you abstract the backend services with a gateway façade that provides  flexibility, control, and security?
```

> Then, how will back-end core services communicate with each other? Will you allow direct HTTP calls that lead to coupling and impact performance and agility? Or, might you consider decoupled messaging with queue and topic technologies?

Reviewer: ```suggestion
How will back-end core services communicate with each other? Will you allow direct HTTP calls that lead to coupling and impact performance and agility? Or might you consider decoupled messaging with queue and topic technologies?
```

*Resiliency*

> A microservices architecture moves your system from in-process to network communication. In a distributed environment, what will you do when Service B isn't responding to a call from Service A? What happens when Service C becomes temporarily unavailable and other services calling it, stack up, and degrade system performance?

Reviewer: ```suggestion
A microservices architecture moves your system from in-process to network communication. In a distributed environment, what will you do when Service B isn't responding to a call from Service A? What happens when Service C becomes temporarily unavailable and other services calling it stack up and degrade system performance?
```
